// @thirdweb-dev/wallets/evm/wallets/aws-secrets-manager/dist/thirdweb-dev-wallets-evm-wallets-aws-secrets-manager.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../../dist/defineProperty-6ca2d9a5.cjs.prod.js');
var evm_wallets_abstract_dist_thirdwebDevWalletsEvmWalletsAbstract = require('../../../../dist/abstract-221871d8.cjs.prod.js');
var clientSecretsManager = require('@aws-sdk/client-secrets-manager');
var ethers = require('ethers');
require('eventemitter3');

/**
 * Create a wallet instance using a private key stored in AWS Secrets Manager.
 *
 *  @example
 * ```javascript
 * import { ThirdwebSDK } from "@thirdweb-dev/sdk"
 * import { AwsSecretsManagerWallet } from "@thirdweb-dev/sdk/evm/wallets"
 *
 * const wallet = new AwsSecretsManagerWallet({
 *   secretName: "my-secret",
 *   secretKeyName: "private-key",
 *   awsConfig: {
 *     region: "us-east-1",
 *     credentials: {
 *       accessKeyId: process.env.AWS_ACCESS_KEY_ID,
 *       secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
 *     },
 *   }
 * });
 *
 * const sdk = await ThirdwebSDK.fromWallet(wallet, "mainnet");
 * ```
 */
class AwsSecretsManagerWallet extends evm_wallets_abstract_dist_thirdwebDevWalletsEvmWalletsAbstract.AbstractWallet {
  constructor(options) {
    super();
    defineProperty._defineProperty(this, "secretId", void 0);
    defineProperty._defineProperty(this, "secretKeyName", void 0);
    defineProperty._defineProperty(this, "client", void 0);
    this.secretId = options.secretId;
    this.secretKeyName = options.secretKeyName;
    this.client = new clientSecretsManager.SecretsManagerClient(options.awsConfig || {});
  }
  async getSigner() {
    const res = await this.client.send(new clientSecretsManager.GetSecretValueCommand({
      SecretId: this.secretId,
      VersionStage: "AWSCURRENT" // VersionStage defaults to AWSCURRENT if unspecified
    }));

    if (!res || !res.SecretString) {
      throw new Error(`No secret found at ${this.secretId}`);
    }
    let privateKey;
    try {
      privateKey = JSON.parse(res.SecretString)[this.secretKeyName];
    } catch {
      throw new Error(`Secret ${this.secretId} is not a valid JSON object! Please convert secret to a JSON object with key ${this.secretKeyName}.`);
    }
    if (!privateKey) {
      throw new Error(`Secret ${this.secretId} does not have key ${this.secretKeyName}`);
    }
    return new ethers.ethers.Wallet(privateKey);
  }
}

exports.AwsSecretsManagerWallet = AwsSecretsManagerWallet;
