// @thirdweb-dev/wallets/evm/wallets/injected/dist/thirdweb-dev-wallets-evm-wallets-injected.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../../dist/defineProperty-21d22449.cjs.dev.js');
var classPrivateFieldGet = require('../../../../dist/classPrivateFieldGet-6e4ddeb2.cjs.dev.js');
var twConnector = require('../../../../dist/tw-connector-8ebaa5b3.cjs.dev.js');
var evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase = require('../../base/dist/thirdweb-dev-wallets-evm-wallets-base.cjs.dev.js');
require('eventemitter3');
require('../../../../dist/abstract-87f231fe.cjs.dev.js');
require('ethers');
require('../../../../dist/storage-b818cf72.cjs.dev.js');
require('localforage');

var _connector = /*#__PURE__*/new WeakMap();
class InjectedWallet extends evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase.AbstractBrowserWallet {
  get walletName() {
    return "Injected Wallet";
  }
  constructor(options) {
    super(InjectedWallet.id, options);
    classPrivateFieldGet._classPrivateFieldInitSpec(this, _connector, {
      writable: true,
      value: void 0
    });
  }
  async getConnector() {
    if (!classPrivateFieldGet._classPrivateFieldGet(this, _connector)) {
      // import the connector dynamically
      const {
        InjectedConnector
      } = await Promise.resolve().then(function () { return require('../../../connectors/injected/dist/thirdweb-dev-wallets-evm-connectors-injected.cjs.dev.js'); });
      classPrivateFieldGet._classPrivateFieldSet(this, _connector, new twConnector.WagmiAdapter(new InjectedConnector({
        chains: this.chains,
        options: {
          shimDisconnect: true
        }
      })));
    }
    return classPrivateFieldGet._classPrivateFieldGet(this, _connector);
  }
}
defineProperty._defineProperty(InjectedWallet, "id", "injected");

exports.InjectedWallet = InjectedWallet;
