// @thirdweb-dev/wallets/evm/connectors/device-wallet/dist/thirdweb-dev-wallets-evm-connectors-device-wallet.esm.js
import { _ as _classPrivateFieldInitSpec, b as _classPrivateFieldSet, a as _classPrivateFieldGet } from '../../../../dist/classPrivateFieldGet-5cbab388.esm.js';
import { _ as _defineProperty } from '../../../../dist/defineProperty-e24c82ea.esm.js';
import { T as TWConnector } from '../../../../dist/tw-connector-e44dbd6f.esm.js';
import { ethers } from 'ethers';
import 'eventemitter3';

var _wallet = /*#__PURE__*/new WeakMap();
var _provider = /*#__PURE__*/new WeakMap();
var _signer = /*#__PURE__*/new WeakMap();
class DeviceWalletConnector extends TWConnector {
  constructor(options) {
    super();
    _defineProperty(this, "id", "device_wallet");
    _defineProperty(this, "name", "Device Wallet");
    _defineProperty(this, "options", void 0);
    _defineProperty(this, "chainId", void 0);
    _classPrivateFieldInitSpec(this, _wallet, {
      writable: true,
      value: void 0
    });
    _classPrivateFieldInitSpec(this, _provider, {
      writable: true,
      value: void 0
    });
    _classPrivateFieldInitSpec(this, _signer, {
      writable: true,
      value: void 0
    });
    _defineProperty(this, "shimDisconnectKey", "deviceWallet.shimDisconnect");
    this.options = options;
    this.chainId = options.chain.chainId;
    _classPrivateFieldSet(this, _wallet, options.wallet);
  }
  async connect(args) {
    if (args.chainId) {
      this.chainId = args.chainId;
    }
    await this.initializeDeviceWallet(args.password);
    const signer = await this.getSigner();
    const address = await signer.getAddress();
    return address;
  }
  async initializeDeviceWallet(password) {
    // TODO this should be a UI flow prior to calling connect instead
    const savedAddr = await _classPrivateFieldGet(this, _wallet).getSavedWalletAddress();
    if (!savedAddr) {
      await _classPrivateFieldGet(this, _wallet).generateNewWallet();
      await _classPrivateFieldGet(this, _wallet).save(password);
    } else {
      await _classPrivateFieldGet(this, _wallet).loadSavedWallet(password);
    }
  }
  async disconnect() {
    _classPrivateFieldSet(this, _provider, undefined);
    _classPrivateFieldSet(this, _signer, undefined);
  }
  async getAddress() {
    const signer = await this.getSigner();
    if (!signer) {
      throw new Error("No signer found");
    }
    return await signer.getAddress();
  }
  async isConnected() {
    try {
      const addr = await this.getAddress();
      return !!addr;
    } catch {
      return false;
    }
  }
  async getChainId() {
    return this.chainId;
  }
  async getProvider() {
    if (!_classPrivateFieldGet(this, _provider)) {
      _classPrivateFieldSet(this, _provider, new ethers.providers.JsonRpcBatchProvider(this.options.chain.rpc[0]));
    }
    return _classPrivateFieldGet(this, _provider);
  }
  async getSigner() {
    if (!_classPrivateFieldGet(this, _wallet)) {
      throw new Error("No wallet found");
    }
    if (!_classPrivateFieldGet(this, _signer)) {
      const provider = await this.getProvider();
      _classPrivateFieldSet(this, _signer, await _classPrivateFieldGet(this, _wallet).getSigner(provider));
    }
    return _classPrivateFieldGet(this, _signer);
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  async switchChain(chainId) {
    // TODO
    throw new Error("Not supported");
  }
}

export { DeviceWalletConnector };
