// @thirdweb-dev/wallets/evm/wallets/email-wallet/dist/thirdweb-dev-wallets-evm-wallets-email-wallet.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../../dist/defineProperty-6ca2d9a5.cjs.prod.js');
var classPrivateFieldGet = require('../../../../dist/classPrivateFieldGet-ee0bcee0.cjs.prod.js');
var evm_connectors_email_dist_thirdwebDevWalletsEvmConnectorsEmail = require('../../../connectors/email/dist/thirdweb-dev-wallets-evm-connectors-email.cjs.prod.js');
var evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase = require('../../base/dist/thirdweb-dev-wallets-evm-wallets-base.cjs.prod.js');
require('../../../../dist/tw-connector-180d18e7.cjs.prod.js');
require('eventemitter3');
require('@paperxyz/embedded-wallet-service-sdk');
require('../../../../dist/abstract-221871d8.cjs.prod.js');
require('ethers');
require('../../../../dist/storage-97213a4f.cjs.prod.js');
require('localforage');

var _connector = /*#__PURE__*/new WeakMap();
class EmailWallet extends evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase.AbstractBrowserWallet {
  get walletName() {
    return "Email Wallet";
  }
  constructor(options) {
    super(EmailWallet.id, {
      ...options,
      shouldAutoConnect: false // TODO figure the autoconnect flow
    });
    classPrivateFieldGet._classPrivateFieldInitSpec(this, _connector, {
      writable: true,
      value: void 0
    });
  }
  async getConnector() {
    if (!classPrivateFieldGet._classPrivateFieldGet(this, _connector)) {
      const chainName = evm_connectors_email_dist_thirdwebDevWalletsEvmConnectorsEmail.PaperChainMap[this.options.chainId];
      if (!chainName) {
        throw new Error("Unsupported chain id: " + this.options.chainId);
      }
      const {
        EmailWalletConnector
      } = await Promise.resolve().then(function () { return require('../../../connectors/email/dist/thirdweb-dev-wallets-evm-connectors-email.cjs.prod.js'); });
      classPrivateFieldGet._classPrivateFieldSet(this, _connector, new EmailWalletConnector({
        clientId: this.options.clientId,
        chain: chainName
      }));
    }
    return classPrivateFieldGet._classPrivateFieldGet(this, _connector);
  }
}
defineProperty._defineProperty(EmailWallet, "id", "email-wallet");

exports.EmailWallet = EmailWallet;
