// @thirdweb-dev/wallets/evm/wallets/email-wallet/dist/thirdweb-dev-wallets-evm-wallets-email-wallet.esm.js
import { _ as _defineProperty } from '../../../../dist/defineProperty-e24c82ea.esm.js';
import { _ as _classPrivateFieldInitSpec, a as _classPrivateFieldGet, b as _classPrivateFieldSet } from '../../../../dist/classPrivateFieldGet-5cbab388.esm.js';
import { PaperChainMap } from '../../../connectors/email/dist/thirdweb-dev-wallets-evm-connectors-email.esm.js';
import { AbstractBrowserWallet } from '../../base/dist/thirdweb-dev-wallets-evm-wallets-base.esm.js';
import '../../../../dist/tw-connector-e44dbd6f.esm.js';
import 'eventemitter3';
import '@paperxyz/embedded-wallet-service-sdk';
import '../../../../dist/abstract-c169745b.esm.js';
import 'ethers';
import '../../../../dist/storage-d861365e.esm.js';
import 'localforage';

var _connector = /*#__PURE__*/new WeakMap();
class EmailWallet extends AbstractBrowserWallet {
  get walletName() {
    return "Email Wallet";
  }
  constructor(options) {
    super(EmailWallet.id, {
      ...options,
      shouldAutoConnect: false // TODO figure the autoconnect flow
    });
    _classPrivateFieldInitSpec(this, _connector, {
      writable: true,
      value: void 0
    });
  }
  async getConnector() {
    if (!_classPrivateFieldGet(this, _connector)) {
      const chainName = PaperChainMap[this.options.chainId];
      if (!chainName) {
        throw new Error("Unsupported chain id: " + this.options.chainId);
      }
      const {
        EmailWalletConnector
      } = await import('../../../connectors/email/dist/thirdweb-dev-wallets-evm-connectors-email.esm.js');
      _classPrivateFieldSet(this, _connector, new EmailWalletConnector({
        clientId: this.options.clientId,
        chain: chainName
      }));
    }
    return _classPrivateFieldGet(this, _connector);
  }
}
_defineProperty(EmailWallet, "id", "email-wallet");

export { EmailWallet };
