// @zag-js/core/dist/types.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/types.ts
var types_exports = {};
__export(types_exports, {
  ActionTypes: () => ActionTypes,
  MachineStatus: () => MachineStatus,
  MachineType: () => MachineType
});
module.exports = __toCommonJS(types_exports);
var MachineStatus = /* @__PURE__ */ ((MachineStatus2) => {
  MachineStatus2["NotStarted"] = "Not Started";
  MachineStatus2["Running"] = "Running";
  MachineStatus2["Stopped"] = "Stopped";
  return MachineStatus2;
})(MachineStatus || {});
var ActionTypes = /* @__PURE__ */ ((ActionTypes2) => {
  ActionTypes2["Start"] = "machine.start";
  ActionTypes2["Stop"] = "machine.stop";
  ActionTypes2["Created"] = "machine.created";
  ActionTypes2["SendParent"] = "machine.send-parent";
  ActionTypes2["After"] = "machine.after";
  ActionTypes2["Every"] = "machine.every";
  ActionTypes2["Init"] = "machine.init";
  return ActionTypes2;
})(ActionTypes || {});
var MachineType = /* @__PURE__ */ ((MachineType2) => {
  MachineType2["Machine"] = "machine";
  MachineType2["Actor"] = "machine.actor";
  return MachineType2;
})(MachineType || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ActionTypes,
  MachineStatus,
  MachineType
});
