// @thirdweb-dev/wallets/evm/connectors/email/dist/thirdweb-dev-wallets-evm-connectors-email.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../../dist/defineProperty-6ca2d9a5.cjs.prod.js');
var twConnector = require('../../../../dist/tw-connector-180d18e7.cjs.prod.js');
var embeddedWalletServiceSdk = require('@paperxyz/embedded-wallet-service-sdk');
require('eventemitter3');

const PaperChainMap = {
  1: "Ethereum",
  5: "Goerli",
  137: "Polygon",
  80001: "Mumbai"
};
class EmailWalletConnector extends twConnector.TWConnector {
  constructor(options) {
    super();
    defineProperty._defineProperty(this, "id", "email-wallet");
    defineProperty._defineProperty(this, "name", "Email Wallet");
    defineProperty._defineProperty(this, "ready", true);
    defineProperty._defineProperty(this, "user", null);
    defineProperty._defineProperty(this, "paper", void 0);
    this.paper = new embeddedWalletServiceSdk.PaperEmbeddedWalletSdk({
      clientId: options.clientId,
      chain: options.chain
    });
  }

  // TODO define our own connector interface
  async connect(args) {
    const email = args?.email;
    if (!email) {
      throw new Error("No Email provided");
    }
    let getUser = await this.paper.getUser();
    switch (getUser.status) {
      case embeddedWalletServiceSdk.UserStatus.LOGGED_OUT:
        {
          const authResult = await this.paper.auth.loginWithPaperEmailOtp({
            email
          });
          this.user = authResult.user;
          break;
        }
      case embeddedWalletServiceSdk.UserStatus.LOGGED_IN_WALLET_INITIALIZED:
        {
          this.user = getUser;
          break;
        }
    }
    if (!this.user) {
      throw new Error("Error connecting User");
    }
    return this.getAddress();
  }
  async disconnect() {
    this.user = null;
  }
  async getAddress() {
    return await this.getUser().wallet.getAddress();
  }
  async isConnected() {
    try {
      const addr = await this.getAddress();
      return !!addr;
    } catch (e) {
      return false;
    }
  }
  async getProvider() {
    const signer = await this.getSigner();
    if (!signer.provider) {
      throw new Error("Provider not found");
    }
    return signer.provider;
  }
  async getSigner() {
    // TODO get RPC from chains package
    const signer = this.user?.wallet.getEthersJsSigner({
      rpcEndpoint: "https://mumbai.rpc.thirdweb.com"
    });
    if (!signer) {
      throw new Error("Signer not found");
    }
    return signer;
  }
  async isAuthorized() {
    return false;
  }
  async switchChain(chainId) {
    const chainName = PaperChainMap[chainId];
    if (!chainName) {
      throw new Error("Chain not supported");
    }
    this.user?.wallet.setChain({
      chain: chainName
    });
  }
  getUser() {
    if (!this.user) {
      throw new Error("User not found");
    }
    return this.user;
  }
}

exports.EmailWalletConnector = EmailWalletConnector;
exports.PaperChainMap = PaperChainMap;
