// @thirdweb-dev/wallets/dist/storage-b818cf72.cjs.dev.js
'use strict';

var localforage = require('localforage');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var localforage__default = /*#__PURE__*/_interopDefault(localforage);

function prefixKey(key) {
  return `tw_${key}`;
}
let connectorStorage;
function getConnectorStorage() {
  if (!connectorStorage) {
    connectorStorage = localforage__default["default"].createInstance({
      name: prefixKey("connectors"),
      storeName: prefixKey("connectors"),
      version: 1.0
    });
  }
  return connectorStorage;
}
let coordinatorStorage;
function getCoordinatorStorage() {
  if (!coordinatorStorage) {
    coordinatorStorage = localforage__default["default"].createInstance({
      name: prefixKey("coordinator"),
      storeName: prefixKey("coordinator"),
      version: 1.0
    });
  }
  return coordinatorStorage;
}
const walletStorage = new Map();
function getWalletStorage(walletId) {
  let storage = walletStorage.get(walletId);
  if (!storage) {
    storage = localforage__default["default"].createInstance({
      name: prefixKey(`wallet:${walletId}`),
      storeName: prefixKey(`wallet_${walletId}`),
      version: 1.0
    });
    walletStorage.set(walletId, storage);
  }
  return storage;
}

exports.getConnectorStorage = getConnectorStorage;
exports.getCoordinatorStorage = getCoordinatorStorage;
exports.getWalletStorage = getWalletStorage;
