// @thirdweb-dev/wallets/dist/tw-connector-8ebaa5b3.cjs.dev.js
'use strict';

var defineProperty = require('./defineProperty-21d22449.cjs.dev.js');
var EventEmitter = require('eventemitter3');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var EventEmitter__default = /*#__PURE__*/_interopDefault(EventEmitter);

class TWConnector extends EventEmitter__default["default"] {}
class WagmiAdapter extends TWConnector {
  constructor(wagmiConnector) {
    super();
    defineProperty._defineProperty(this, "wagmiConnector", void 0);
    this.wagmiConnector = wagmiConnector;
  }
  async connect(args) {
    const chainId = args?.chainId;
    const result = await this.wagmiConnector.connect({
      chainId
    });
    return result.account;
  }
  disconnect() {
    return this.wagmiConnector.disconnect();
  }
  isConnected() {
    return this.wagmiConnector.isAuthorized();
  }
  getAddress() {
    return this.wagmiConnector.getAccount();
  }
  getSigner() {
    return this.wagmiConnector.getSigner();
  }
  getProvider() {
    return this.wagmiConnector.getProvider();
  }
  async switchChain(chainId) {
    if (!this.wagmiConnector.switchChain) {
      throw new Error("Switch chain not supported");
    }
    await this.wagmiConnector.switchChain(chainId);
  }
}

exports.TWConnector = TWConnector;
exports.WagmiAdapter = WagmiAdapter;
