// @thirdweb-dev/wallets/evm/wallets/metamask/dist/thirdweb-dev-wallets-evm-wallets-metamask.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../../dist/defineProperty-6ca2d9a5.cjs.prod.js');
var classPrivateFieldGet = require('../../../../dist/classPrivateFieldGet-ee0bcee0.cjs.prod.js');
var twConnector = require('../../../../dist/tw-connector-180d18e7.cjs.prod.js');
var evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase = require('../../base/dist/thirdweb-dev-wallets-evm-wallets-base.cjs.prod.js');
require('eventemitter3');
require('../../../../dist/abstract-221871d8.cjs.prod.js');
require('ethers');
require('../../../../dist/storage-97213a4f.cjs.prod.js');
require('localforage');

var _connector = /*#__PURE__*/new WeakMap();
class MetaMask extends evm_wallets_base_dist_thirdwebDevWalletsEvmWalletsBase.AbstractBrowserWallet {
  get walletName() {
    return "MetaMask";
  }
  constructor(options) {
    super(MetaMask.id, options);
    classPrivateFieldGet._classPrivateFieldInitSpec(this, _connector, {
      writable: true,
      value: void 0
    });
  }
  async getConnector() {
    if (!classPrivateFieldGet._classPrivateFieldGet(this, _connector)) {
      // import the connector dynamically
      const {
        MetaMaskConnector
      } = await Promise.resolve().then(function () { return require('../../../connectors/metamask/dist/thirdweb-dev-wallets-evm-connectors-metamask.cjs.prod.js'); });
      classPrivateFieldGet._classPrivateFieldSet(this, _connector, new twConnector.WagmiAdapter(new MetaMaskConnector({
        chains: this.chains,
        options: {
          shimDisconnect: true
        }
      })));
    }
    return classPrivateFieldGet._classPrivateFieldGet(this, _connector);
  }
}
defineProperty._defineProperty(MetaMask, "id", "metamask");

exports.MetaMask = MetaMask;
